app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt

```diff
@@ -102,10 +100,16 @@ class HomeFragment : Fragment(), CoroutineScope {
 
         setupHomeMenu()
 
-        val searchIcon = requireComponents.search.searchEngineManager.getDefaultSearchEngine(
-            requireContext()
-        ).let {
-            BitmapDrawable(resources, it.icon)
+        GlobalScope.async{
```

Comment:
GlobalScope is generally frowned upon. That's especially the case outside of Application.onCreate.

"The reason to avoid GlobalScope" by Roman Elizarov... Roman is the guy who made Kotlin Coroutines.
https://link.medium.com/egfyXYK6fW

Also, async will crash on an exception. In this case, we'll crash if the network connection fails. Launch seems more appropriate.

Reply:
Thanks for the feedback. I am correcting it now and I will resubmit!

Reply:
I read this article and then started to address your comments by fixing the code. Unfortunately, I think that that the article is talking about a situation that is very different than the one that I am working on here. In the article, the author is specifically talking about invoking several operations that should run in parallel and waiting for all of them to complete before continuing.

In this case, that is ***not*** what I am trying to do. In fact, that would defeat the whole purpose of this optimization. I am attempting to launch most of the things that start has to do into the background and let the function return as quickly as possible. 

I essentially do the "wait" that he discusses in the `stop` function by waiting for the startup tasks to complete before destroying them. I understand his point about this operation being canceled if the parent is cancelled before the child is finished operating but I do not think that is a risk here because the `start` caller is the main thread and we would want these operations cancelled if the user closes the application.

Does this make sense? Is it reasonable? I would assume that I am missing the point of his article and your comment? Can you redirect me?

Your point about `launch` vs `async` w.r.t the potential problem of a network exception is well taken and I will absolutely correct that immediately.

Reply:
There are actually several issues here, but I attempted to link to an article I couldn't remember anymore to explain why the behavior is frowned on. It seems Roman didn't appropriately explain.

Downloading and showing the image is no longer useful after the user clicks on any button that causes us to move to another Fragment. We already have an appropriate coroutineScope implemented on HomeFragment with the proper cancellation setup to avoid any wasted work if the user clicks away. Using GlobalScope explicitly instead of the Fragment's coroutineScope implicitly means we lose the benefit of this cooperative cancellation. We expend resources loading and setting the icon after the Fragment's work has been cancelled.

Most bugs on Android are caused by race conditions around the platform's complex lifecycle changes. In this case, we're calling GlobalScope.async without a CoroutineContext parameter. Inside the async method, we're setting properties on the views. If the property changes fail and crash the app, then we succeeded to run on a background thread. Background threads are disallowed from making view changes. If the view property changes succeed, then we were actually running on the main thread and blocking the app the whole time in spite of the async {} keyword. In essence, we got no benefit from the coroutine at all.

This all means that it really does matter if we have cooperative cancellation. We need to perform background work on a background thread and UI work on the main thread because Android requires it. If we don't cancel the operation, we'll attempt to access UI views which are no longer onscreen and also crash the app referencing nonexistent views. Using the Fragment's coroutineScope fixes this because of cooperative cancellation, but be sure to specify a background coroutineContext parameter and launch any UI changes onto the Main dispatcher.

It may be okay to use GlobalScope for GleanMetricsService provided that we do not need any child or parent coroutines to cancel cooperatively. I still prefer to create a scope in most cases, since we don't know what later programmers will do in this service.

Reply:
With respect to the use of async in the HomeFragment, I understand its CoroutineScope is Main which means that if I simply `launch` from within any member function, the work will the done on the UI thread. This means that any work within the coroutine can access the UI without fear. That said, it does not do the work on a background thread which means that the creation of the Fragment will not wait for the completion of that work, but it does not take advantage of the possibility that other background threads exist that are waiting to do work.

So, it is ideal to `launch` most of the work in `Dispatchers.Default` and then take care to do the UI updates in the `Dispatchers.Main` context. 

Do I have that correct? This is the way that I am going to move forward on a revision and I will submit it here.

Thanks again for the feedback!

Reply:
I am resolving this conversation based on the version of the PR that I just posted. I hope that I tacked closer to the ideal!